### PR TITLE
Update spi.py

### PR DESCRIPTION
**Pull request by c235labtop**
Remove sleep statement in read_spi_fifo to speed up SW readout to match FW speed

### CODE REVIEW

**Comment by nic-str** [comment]
To improve the performance it is better to readout the fifo in 1KB or 4KB or even larger chunks.

I also added the option to set a maximum number of readcycles to avoid hanging.
https://github.com/nic-str/astropix-python/blob/92e11a14fd567d43965766760ddb2e686c3f8576/modules/spi.py#L232

To make reading more robust, I would also implement this change to the ftdi read function, which assures that you really get the number of bytes back that you want to read
https://github.com/nic-str/astropix-python/blob/d6e15e3bb141cc0e87092bb91f8eb13106427130/modules/nexysio.py#L152

Edit: updated permalinks

**Comment by c235labtop** [comment]
Pending confirmation check on GSFC bench

File: core/nexysio.py [hunk 1]
```diff
@@ -155,8 +155,16 @@ def read(self, num: int) -> bytes:
 
         :param num: Number of Bytes to read
         """
+        remaining = num
+        bytes = bytearray()
+
         try:
-            return self._handle.read(num)
+            while remaining > 0:
+                rbytes = self._handle.read(num)
```
Comment (nic-str) [inline core/nexysio.py]: This line is incorrect, should be:
rbytes = self._handle.read(remaining)

File: core/spi.py [hunk 2]
```diff
@@ -220,17 +229,18 @@ def read_spi_readoutmode(self):
...
+        :returns: SPI read stream
+        """
         read_stream = bytearray()
+        readcount = 0
 
-        while not self.get_spi_config() & 16:
-            readbuffer = self.read_spi(64)
-
+        while not (self.get_spi_config() & SPI_READ_FIFO_EMPTY) and readcount<max_reads:
+            readbuffer = self.read_spi(1024)
```
Comment (nic-str) [inline core/spi.py]: I did some further testing, increasing this to 4096 resulted in a significant performance improvement.
With a simple testscript with just one loop for readout, decoding and writing to a file, I can process 70k Hits per second.

```python
while time.time() < stop_time:
        readout = nexys.read_spi_fifo(10)

        if readout != b'':
            cnt += 1
            print(cnt)

            list_hits = decode.hits_from_readoutstream(readout)
            cnt_list_hits += len(list_hits)
            print(cnt_list_hits)
            df = decode.decode_astropix2_hits(list_hits)
  
            df.to_csv(output_path, mode='a', header=not os.path.exists(output_path))
```

**Comment by asteinhebel** [comment]
@nic-str please double check that this covers your comments here - no additional updates from your other work integrated here yet (will make another branch)

**Comment by nic-str** [comment]
> @nic-str please double check that this covers your comments here - no additional updates from your other work integrated here yet (will make another branch)

Looks good to me!

**Comment by asteinhebel** [comment]
Thanks :)

**Comment by c235labtop** [comment]
Confirmed on GSFC bench

**Review by asteinhebel — APPROVED** [review]
Confirmed on GSFC bench

**Pull request merged**